src/modules/bppp/main_impl.h

```diff
@@ -49,67 +60,147 @@ secp256k1_bppp_generators *secp256k1_bppp_generators_create(const secp256k1_cont
     return ret;
 }
 
-secp256k1_bppp_generators* secp256k1_bppp_generators_parse(const secp256k1_context* ctx, const unsigned char* data, size_t data_len) {
-    size_t n = data_len / 33;
-    secp256k1_bppp_generators* ret;
-
+void secp256k1_bppp_generators_destroy(const secp256k1_context* ctx, secp256k1_bppp_generators *gens) {
     VERIFY_CHECK(ctx != NULL);
-    ARG_CHECK(data != NULL);
-
-    if (data_len % 33 != 0) {
-        return NULL;
+    (void) ctx;
+    if (gens != NULL) {
+        free(gens->gens);
+        free(gens);
     }
+}
 
-    ret = (secp256k1_bppp_generators *)checked_malloc(&ctx->error_callback, sizeof(*ret));
-    if (ret == NULL) {
-        return NULL;
-    }
-    ret->n = n;
-    ret->gens = (secp256k1_ge*)checked_malloc(&ctx->error_callback, n * sizeof(*ret->gens));
-    if (ret->gens == NULL) {
-        free(ret);
-        return NULL;
+size_t secp256k1_bppp_rangeproof_proof_length(
+    const secp256k1_context* ctx,
+    size_t n_bits,
+    size_t base
+) {
+    size_t num_digits, n_rounds, g_len, h_len;
+    VERIFY_CHECK(ctx != NULL);
+    if (n_bits > 64 || base < 2 || base > 64) {
+        return 0;
     }
 
-    while (n--) {
-        secp256k1_generator gen;
-        if (!secp256k1_generator_parse(ctx, &gen, &data[33 * n])) {
-            free(ret->gens);
-            free(ret);
-            return NULL;
-        }
-        secp256k1_generator_load(&ret->gens[n], &gen);
+    if (!secp256k1_is_power_of_two(base) || !secp256k1_is_power_of_two(n_bits)) {
+        return 0;
     }
-    return ret;
+    num_digits = n_bits / secp256k1_bppp_log2(base);
+    if (!secp256k1_is_power_of_two(num_digits)) {
+        return 0;
+    }
+    g_len = num_digits > base ? num_digits : base;
+    h_len = 8;
+    n_rounds = secp256k1_bppp_log2(g_len > h_len ? g_len : h_len);
+    return 33 * 4 + 65*n_rounds + 64;
```

Comment:
@tevador, this does not implement all the optimizations with early terminations.

Reply:
Searching for "early termination" in the paper gives nothing. Can you point me to the relevant section of the paper?

Reply:
OK, I think I figured it out from the code. If you do just 3 rounds instead of 4, you'll end up with `n_vec` of length 2 and `l_vec` of length 1, which is actually more compact.

Reply:
In the paper, we stop the recursion already when `len(l_vec) + len(n_vec) <= 6`, see page 13.

Reply:
I think it should be `len(l_vec) + len(n_vec) < 6`. After 2 rounds, we get `len(l_vec) = 2` and `len(n_vec) = 4`, but the proof can still be reduced by 31 or 32 bytes by performing the 3rd round.